ARROW-2361: [Rust] Starting point for a native Rust implementation of Arrow

andygrove opened the pull request:
This PR implements a subset of data types and array functionality, supporting primitives, strings, and structs.

I am actively developing this code and relying on it in my DataFusion project.

It isn't fully Arrow compatible yet but I think this is a solid foundation to build from.

jihoonson commented on rust/src/array.rs:
```diff
@@ -0,0 +1,248 @@
+
+use bytes::{Bytes, BytesMut, BufMut};
+
+pub enum ArrayData {
```
Probably you have a plan to make this immutable and add a sort of `ArrayBuilder` in the future. It would be great if it can be commented somewhere.

kszucs replied:
`Bytes` have a nice primitive for that, see [freeze](https://carllerche.github.io/bytes/bytes/struct.BytesMut.html#examples).

kszucs commented:
Just for the record We've experienced with a generic [implementation](https://github.com/kszucs/arrow/blob/list/rust/src/array.rs), but as well as the `enum` based approach it has its upsides and downsides. 

It'd provide zero cost abstraction, better extensibility, better error handling, less repetition. 
But due to the fact, that it does everything at compile time, runtime construction of `Table(Vec<Box<Array<T...>>)`, `StructType(<Vec<(String, Box<DataType>))` and Unions types/arrays would require runtime reflection, mostly achieveable by `downcasting`, which makes it more complicated.

> However with Andy We've concluded that the `enum` way is easier to start with, I'll still experiment with the generic implementation (if anyone interested), especially because it works well at compile time and I really like the following syntax:

```rust
#[derive(Arrow)]
struct MyType {
    a: Int64.
    b: String
}

let mut arr = Array::<MyType>::empty();
```
> Also speed, speed, speed :)

kszucs commented on rust/src/array.rs:
```diff
@@ -0,0 +1,248 @@
+pub struct ListData {
+    pub offsets: Vec<i32>,
+    pub bytes: Bytes
+}
```
How about placing it in the `ArrayData` enum, something like [here](https://github.com/kszucs/arrow/blob/enum/rust/src/e.rs#L66)?

kszucs commented on rust/src/array.rs:
```diff
@@ -0,0 +1,248 @@
+
+use super::bitmap::Bitmap;
+
+use bytes::{Bytes, BytesMut, BufMut};
```
I think we should go with [RawVec](https://doc.rust-lang.org/alloc/raw_vec/struct.RawVec.html), at least for now. 
Provides pretty nice abstraction, see [List push](https://github.com/kszucs/arrow/blob/enum/rust/src/array.rs#L91), and we'll be able to implement memory alignment sooner with the `allocator_api`.

andygrove replied:
The issue is that then we can't use Rust stable and would require nightly

andygrove replied:
For memory allocation I think we should do what @jihoonson did in iron-arrow and use libc since it lets us have aligned memory without requiring nightly rust.

andygrove commented:
The build is failing with `NOT APPROVED: rust/README.md (apache-arrow/rust/README.md): false` ... anyone know what I need to do about that?

kszucs commented:
It's https://creadur.apache.org/rat/ I guess. Misses license note from README? Probably need to whitelist somewhere.

wesm commented:
Don't exclude. Look at other README.md files in the codebase -- they have copyright headers.

andygrove commented:
@jihoonson @kszucs I added some TODO comments about the questions and issues you raised. I think if we can get this initial PR merged then we can start building very quickly with additional PRs.

wesm commented on rust/Cargo.toml:
```diff
@@ -21,7 +21,7 @@ version = "0.1.0"
 homepage = "https://github.com/apache/arrow"
 repository = "https://github.com/apache/arrow"
-authors = ["Andy Grove <andygrove73@gmail.com>"]
+authors = ["Andy Grove <andygrove73@gmail.com>", "Krisztián Szűcs <szucs.krisztian@gmail.com>"]
```
This should be `dev@arrow.apache.org`

wesm commented:
I'm sorta ambivalent on the package name -- I looked at crates.io and there are some other ASF projects with packages that just use the Foo in Apache Foo. If "arrow" is shorter and sweeter, that's no problem

andygrove commented:
I've converted the code to use macros to add consistent functionality for all the primitive types that are supported so far. The code is much cleaner now. Also, arrays are now correctly memory-aligned and immutable.

I'm thinking this is good point for another review and maybe merge this PR?

xhochy commented:
@pacman82 This may be of interest for you ;)

andygrove commented:
I just implemented the ability to compare two same-typed arrays using closures. For example:

```
    fn test_array_eq() {
        let a = Array::from(vec![1,2,3,4,5]);
        let b = Array::from(vec![5,4,3,2,1]);
        let c = a.compare_array(&b, &|a: i32,b: i32| a == b).unwrap();
        assert_eq!(c, vec![false,false,true,false,false]);
    }
```

wesm commented:
If there are no objections I can go ahead and merge this so subsequent work can happen in new PRs. I'll wait an hour or so from now before merging

fsaintjacques commented:
I'd like to see the `Buffer` concept from C++ ported here. The main reason is `mmap`. I'd also refactor the `Bitmap` struct to use a Buffer.

fsaintjacques commented on rust/src/array.rs:
```diff
@@ -0,0 +1,412 @@
+    UInt32(*const u32),
+    UInt64(*const u64),
+    Utf8(ListData),
+    Struct(Vec<Rc<Array>>)
```
Should be `Field` instead of `Array`.

andygrove replied:
We have Schema/Field for defining types including structs (I was wrong about this in our slack chat earlier). I added a test with an example for defining a struct.

```
        let _person = Schema::new(vec![
            Field::new("first_name", DataType::Utf8, false),
            Field::new("last_name", DataType::Utf8, false),
            Field::new("address", DataType::Struct(vec![
                Field::new("street", DataType::Utf8, false),
                Field::new("zip", DataType::UInt16, false),
            ]), false),
        ]);
```

andygrove commented:
Hi @fsaintjacques .. thanks for the review. I have added type-safe Buffer<T> type and updated ArrayData to use that. Also Bitmap now uses Buffer too.

andygrove commented:
I have now also re-implemented List support so that there is a type-safe `List<T>` which uses `Buffer<T>` and `ArrayData::Utf8` now is just a wrapper for `List<u8>`.

@fsaintjacques I think this is ready for merge now?
